Tests: Add ability to generate random new fields for xContent parsing test

cbuescher opened the pull request:
For the response parsing we want to be lenient when it comes to parsing new
xContent fields. In order to ensure this in our testing, this change adds a
utility method to ESTestCase that takes xContent bytes representation as input
and recursively adds random fields on each object level.

Some exceptions to consider: sometimes we want to exclude a whole subtree from
this treatment (e.g. skipping "_source"), other times an element (e.g. "fields",
"highlight" in SearchHit) can have arbitraryly named objects. Those cases can be
specified as exceptions.

Three use cases (SearchHitsTests, SearchHitTest, MainResponseTests) are also
included.

cbuescher committed:
Tests: Add ability to generate random new fields for xContent test

For the response parsing we want to be lenient when it comes to parsing new
xContent fields. In order to ensure this in our testing, this change adds a
utility method to ESTestCase that takes xContent bytes representation as input
and recursively adds random fields on each object level.

Some expections to consider: sometimes we want to exclude a whole subtree from
this treatment (e.g. skipping "_source"), other times an element (e.g. "fields",
"highlight" in SearchHit) can have arbitraryly named objects. Those cases can be
specified as exceptions.

Three use cases (SearchHitsTests, SearchHitTest, MainResponseTests) are also
included.

cbuescher commented:
Labeld WIP since it is a little bit rough on the edges but maybe serves as basis for discussion.

cbuescher commented on core/src/main/java/org/elasticsearch/search/SearchHit.java:
```diff
@@ -613,7 +614,10 @@ private static void declareMetaDataFields(ObjectParser<Map<String, Object>, Void
+            String name = parser.currentName();
+            ensureExpectedToken(Token.START_OBJECT, parser.nextToken(), parser::getTokenLocation);
+            ensureFieldName(parser, parser.nextToken(), SearchHits.Fields.HITS);
+            innerHits.put(name, SearchHits.fromXContent(parser));
```
This needed to change because tests uncovered that we didn't correctly parse this structure. It looks like this:
```
"inner_hits": {
           "<inner_hits_name>": {
              "**hits**": {
                 "total": ...,
                 "hits": [
                    {
                       "_type": ...,
                       "_id": ...,
                       ...
                    },
                    ...
                 ]
              }
           }
        },
        ...
```
but it turns out we somehow gracefully read over the first `hits` level and parsed the rest correctly, but this doesn't work any more when there are random bits in between.

cbuescher committed:
Using ObjectPath

tlrx commented on test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:
```diff
@@ -913,6 +909,76 @@ protected final BytesReference toShuffledXContent(ToXContent toXContent, XConten
+     * }
+     * </pre>
+     */
+    public BytesReference insertRandomFields(XContentType contentType, BytesReference xContent, Predicate<String> excludeFilter)
```
EsTestCase has already many methods, maybe this one could be located in XContentTestUtils too?

cbuescher replied:
I put it here to make it easier accessible for tests and also because if I move this to XContentTestUtils, we would need to pass in the RandomizedContext of the test every time. I don't know if that makes up for another method in ESTestCase that we might use in several places? Wdyt?

tlrx replied:
Ok - I don't have a strong feeling about it, we already have shuffledXContent & Co. Let's leave it here then.

tlrx commented on test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:
```diff
@@ -913,6 +909,76 @@ protected final BytesReference toShuffledXContent(ToXContent toXContent, XConten
+        }
+        try (XContentParser parser = createParser(contentType.xContent(), xContent)) {
+            List<String> possiblePaths = XContentTestUtils.getInsertPaths(parser);
+            possiblePaths.stream().filter(excludeFilter.negate()).forEach(insertPaths::add);
```
I think we do this streamed filtering only if excludeFilter is not null

cbuescher replied:
Good point, I will change.

tlrx commented on test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:
```diff
@@ -913,6 +909,76 @@ protected final BytesReference toShuffledXContent(ToXContent toXContent, XConten
+        try (XContentParser parser = createParser(contentType.xContent(), xContent)) {
+            Supplier<Object> value = () -> {
+                if (randomBoolean()) {
+                    return randomAlphaOfLength(10);
```
Should we generate variable length content? And mix with numbers?

cbuescher replied:
Yes, I think I can use RandomObjects.randomStoredFieldValues() which gives us Strings, all kinds of numbers, boolean or a byte[]. Maybe this will make test failures a bit harder to read but since we expect parsers to skip these values, any randomization here should work (in theory).

tlrx reviewed (approved):
Thanks @cbuescher and sorry for the time it took me to review this.

I like how you made the whole thing easy compared to the first version of the pull request. I'm not sure if we really need to add new test methods like testFromXContentLenientParsing or we could simply update the existing testFromXContent methods so that new fields are sometimes inserted. Anyway, I think this should go in so we can try it on aggregations.

cbuescher commented on test/framework/src/main/java/org/elasticsearch/test/XContentTestUtils.java:
```diff
@@ -123,4 +129,54 @@ private static String differenceBetweenObjectsIgnoringArrayOrder(String path, Ob
         }
     }
 
+    public static List<String> getInsertPaths(XContentParser parser) throws IOException{
```
Sure, I will add some explanation of what it does.

cbuescher committed:
Adressing review comments

cbuescher commented:
@tlrx thanks for the review, I added a commit with some changes. Regarding adding additional "testFromXContentLenientParsing()" methods to the test: I added those as extra test methods so when those fail but the "simple" fromXContent test passes, we know that the problem lies in those additional fields we add there. We could randomize it, but I like to have on "basic" test without all possible ranomizations with each test run and one for the additional "complications", at least for the test that check the whole SearchHit(s) output. For the smaller xContent structures I think adding fields randomly in testFromXContent is enough.

tlrx commented:
Thanks @cbuescher, it makes sense.

javanna commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ObjectPath.java:
```diff
@@ -148,4 +149,14 @@ private Object evaluate(String key, Object object, Stash stash) throws IOExcepti
         return list.toArray(new String[list.size()]);
     }
+
+    public XContentBuilder toXContentBuilder(XContent xContent) throws IOException {
```
is this used in one single place? it seems specialized to what we need in insertIntoXContent, maybe it should be part of that method?

cbuescher replied:
Yes, I will move it.

cbuescher replied:
I see why I needed to put this method here. It needs access to the wrapped  `object` field in ObjectPath. I could make that accessible via a getter, but I think I like this better, not exposing the innards of ObjectPath.

javanna replied:
oh right, leave it here then, it's fine. How about making ObjectPath implement ToXContentObject at this point?

javanna commented on test/framework/src/test/java/org/elasticsearch/test/XContentTestUtilsTests.java:
```diff
@@ -0,0 +1,106 @@
+
+public class XContentTestUtilsTests extends ESTestCase {
+
+    public void testGetInserPaths() throws IOException {
```
inser**t**Paths

javanna commented on test/framework/src/test/java/org/elasticsearch/test/XContentTestUtilsTests.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    public void testInsertInto() throws IOException {
```
testInsertIntoXContent?

javanna commented on test/framework/src/main/java/org/elasticsearch/test/XContentTestUtils.java:
```diff
@@ -123,4 +129,84 @@ private static String differenceBetweenObjectsIgnoringArrayOrder(String path, Ob
+     *  <li>"foo3.foo4</li>
+     * </ul>
+     */
+    public static List<String> getInsertPaths(XContentParser parser) throws IOException{
```
these new methods could be package private as they are only called from ESTestCase? Does it make sense to adjust their visibility? After all they are not meant to be used separately, the entry-point should always be insertRandomFields ?

javanna replied:
it wouldn't be bad to actually have them all in the same class though.  Thoughts?

cbuescher replied:
I thought they might be useful as general utility methods for other usecases as well later, although you are right they are quiet specific. They are stateless helpers, so I don't see much harm in leaving them public. About having them in one class: that would mean either moving all of this to ESTestCase or everything to this class, which means having to pass in the random context to `insertRandomFields()` with every call. The separation into static helper classes in this utility class and the test helper that needs access to the tests random context in ESTestCase makes sense to me.

javanna replied:
I see, my thinking is that these methods are quite complicated, naming is hard, and they are better hidden behind their caller. Not a great deal, I just think that we should expose the least possible methods with clear contracts. I do see how this is test only code, but still. Maybe I would reconsider moving the main method to XContentTestUtils. Passing in the random context is not a huge deal maybe?

javanna reviewed (changes requested):
I left a few comments, nothing major. I wonder if this could also be used on day in AbstractQueryTestCase and friends where we inject random fields. Maybe worth having a look, not urgent though and doesn't need to be part of this PR, unless it just works as-is :)

cbuescher commented on core/src/test/java/org/elasticsearch/search/SearchHitTests.java:
```diff
@@ -152,6 +152,33 @@ public void testFromXContent() throws IOException {
+     * everything under "inner_hits". They are also keyed by arbitrary names and contain SearchHits,
+     * which are already tested elsewhere.
+     */
+    public void testFromXContentLenientParsing() throws IOException {
```
I added those as extra test methods so when they fail and the "simple" fromXContent test passes, we know that the problem lies in adding those additional fields. We could randomize it in the existing method, but I like to have one "basic" test without all those extra cases to run with each build as a baseline.  and one for the additional more complicated checks.

javanna replied:
ok makes sense

cbuescher commented on test/framework/src/main/java/org/elasticsearch/test/XContentTestUtils.java:
```diff
@@ -123,4 +129,84 @@ private static String differenceBetweenObjectsIgnoringArrayOrder(String path, Ob
+    public static List<String> getInsertPaths(XContentParser parser) throws IOException{
+        if (parser.currentToken() == null) {
+            parser.nextToken();
+        }
```
You are right, I will change this. I think I added this when I though we might use this from sub-structures as well. I think we can makes this a necessary condition.

cbuescher commented on test/framework/src/main/java/org/elasticsearch/test/XContentTestUtils.java:
```diff
@@ -123,4 +129,84 @@ private static String differenceBetweenObjectsIgnoringArrayOrder(String path, Ob
+    }
+
+    private static List<String> getInsertPaths(XContentParser parser, Stack<String> currentPath) throws IOException {
+        assert (parser.currentToken() == XContentParser.Token.START_OBJECT || parser.currentToken() == XContentParser.Token.START_ARRAY);
```
Yes

cbuescher commented on test/framework/src/main/java/org/elasticsearch/test/XContentTestUtils.java:
```diff
@@ -123,4 +129,84 @@ private static String differenceBetweenObjectsIgnoringArrayOrder(String path, Ob
+        assert (parser.currentToken() == XContentParser.Token.START_OBJECT || parser.currentToken() == XContentParser.Token.START_ARRAY);
+        List<String> validPaths = new ArrayList<>();
+        if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+            if (parser.currentName() != null) {
```
Yes, I will try to do that.

cbuescher replied:
The differentiation between START_OBJECT and START_ARRAY is necessary because the later needs to maintain an item count instead of the current field name to add to the path.

javanna replied:
yea I got that, I didn't mean to suggest to remove that distinction. I only wanted to get rid of the null check against currentName, not sure if I am missing something

cbuescher replied:
I forgot why I needed to do this: when entering objects in arrays, like in `hits: [ {....}, {....}, {....}   ]`, parser.currentNmae() is also null. So this is needed not only for the root object. Took me some time to rework this until I remembered, but I think I struggled with this for a while and the current way was the easierst solution I found.

javanna replied:
yes that is what I was missing, it is not only the root object that needs this, also anonymous objects in arrays :) you do need this check. to make when you need it more explicit you could move the check in the start_array section, and before the start object check for the root object.

cbuescher commented on test/framework/src/main/java/org/elasticsearch/test/XContentTestUtils.java:
```diff
@@ -123,4 +129,84 @@ private static String differenceBetweenObjectsIgnoringArrayOrder(String path, Ob
+        return validPaths;
+    }
+
+    public static XContentBuilder insertIntoXContent(XContent xContent, BytesReference original, List<String> paths, Supplier<String> key,
```
Done

cbuescher committed:
Adderssing more review comments

cbuescher commented:
@javanna thanks for the reviw, I added a commit addressing most of your comments. I will add another one shortly trying to move the whole `ESTestCase#insertRandomFields` method to the XContentTestUtils like you and @tlrx suggested, but this gets even more awkward in terms of using it because we not only need to pass the random context but also the xContentRegistry from the test. While this is still doable, I think I don't like it but I will add another commit shortly so you can judge what you prefer.

cbuescher committed:
Moving insertRandomFields from ESTestCase to XContentTestUtils

cbuescher commented:
@javanna @tlrx I moved the main test method from ESTestCase to XContentTestUtils in 62d6364 but I'm not sure I like it better, there's a lot of stuff that we need to pass in. Any opinions and suggestions how to make this better are welcome, I'm still leaning to keep this in ESTestCase, wdyt?

javanna commented on test/framework/src/main/java/org/elasticsearch/test/XContentTestUtils.java:
```diff
@@ -123,4 +136,163 @@ private static String differenceBetweenObjectsIgnoringArrayOrder(String path, Ob
+     * </pre>
+     */
+    public static BytesReference insertRandomFields(XContentType contentType, BytesReference xContent, Predicate<String> excludeFilter,
+            Random random, NamedXContentRegistry registry) throws IOException {
```
I don't think we need the registry, just make it the empty one and remove the argument?

cbuescher replied:
Good point, will do.

javanna commented on test/framework/src/main/java/org/elasticsearch/test/XContentTestUtils.java:
```diff
@@ -123,4 +136,163 @@ private static String differenceBetweenObjectsIgnoringArrayOrder(String path, Ob
+     *  <li>"foo3.foo4</li>
+     * </ul>
+     */
+    static List<String> getInsertPaths(XContentParser parser) throws IOException{
```
we could potentially remove this method and create the stack where we currently call this method? I think then we don't even need the assertions?

cbuescher replied:
Makes sense after the recent changes.

javanna commented on test/framework/src/main/java/org/elasticsearch/test/XContentTestUtils.java:
```diff
@@ -123,4 +136,163 @@ private static String differenceBetweenObjectsIgnoringArrayOrder(String path, Ob
+        } else if (parser.currentToken() == XContentParser.Token.START_ARRAY) {
+            int itemCount = 0;
+            while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                if (parser.currentToken() == XContentParser.Token.START_OBJECT
```
can we really have a start_array right after a start_array?

javanna replied:
scratch that, we are within a while loop :)

javanna replied:
not sure if worth it, but maybe we could have methods that specifically handle arrays and objects, which can be called directly rather than having to check again the current token. Maybe this complicates things too much. It may help also with the currentName null check, as it can be moved to only the array method.

cbuescher replied:
I tried this already an I didn't like it. It adds other problems too.

javanna commented on test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ObjectPath.java:
```diff
@@ -148,4 +149,20 @@ private Object evaluate(String key, Object object, Stash stash) throws IOExcepti
+        if (this.object instanceof Map) {
+            builder.map((Map<String, Object>) this.object);
+        } else {
+            throw new UnsupportedOperationException("Only ObjectPath created from a map supported.");
```
not that important but maybe we can also support this case, not extremely sure if it works but did you try this?

```
builder.startArray(); 
builder.value((List<?>) this.object);
builder.endArray();
```

cbuescher replied:
I don't think we should to that here and now when we don't need it. ObjectPath is whacky already since it allows to wrap any arbitrary kind of Object, not only lists and maps (I don't think it does anything useful with other kinds of Object, but nethertheless...) I don't think "faking" it can produce any meaningful xContent if the underlying Object is not a Map<String, Object> stretches things a bit.

tlrx commented:
> Any opinions and suggestions how to make this better are welcome, I'm still leaning to keep this in ESTestCase, wdyt?

With @javanna latest suggestions I think I'd prefer to have the method in XContentTestUtils rather than ESTestCase.

cbuescher committed:
Another round on the review carousel

javanna commented on core/src/test/java/org/elasticsearch/action/main/MainResponseTests.java:
```diff
@@ -55,8 +56,10 @@ public void testFromXContent() throws IOException {
         boolean humanReadable = randomBoolean();
         BytesReference originalBytes = toShuffledXContent(mainResponse, xContentType, ToXContent.EMPTY_PARAMS, humanReadable);
+        // we add a few random fields to check that parser is lenient on new fields
+        BytesReference withRandomFields = insertRandomFields(xContentType, originalBytes, null, random());
```
new question :) shall we have then a separate test here too?

cbuescher replied:
In this case I was going with only one test, the xContent structure we test here is not that deep. My thinking was to do this kind of tests for smaller xContent objects and the two separate tests for deeper, top level tests (like SearchHit) where randomization already adds a lot of variety. I will keep this as is for now but will keep an eye on how this goes as we use this in other tests.

javanna commented on test/framework/src/test/java/org/elasticsearch/test/test/ESTestCaseTests.java:
```diff
@@ -159,4 +167,80 @@ public void testRandomUniqueTotallyUnique() {
     }
+
+    @SuppressWarnings("unchecked")
+    public void testInsertRandomXContent() throws IOException {
```
this one needs to be moved to XCOntentTestUtilstTEsts?

cbuescher replied:
Sure.

javanna reviewed (approved):
left two minors, LGTM though, no need for another review round

cbuescher committed:
Moving tests
